test/Feature/HLSLLib/reflect.32.test

```diff
@@ -0,0 +1,209 @@
+#--- source.hlsl
+
+// This test tests three different reflection scenarios:
+// -----> |
+// -----> /
+// -----> -
+// each test case will cycle through 1-4 dimensions.
+
+StructuredBuffer<float4> IncidentRay : register(t0);
+StructuredBuffer<float4> Wall1 : register(t1);
+StructuredBuffer<float4> Wall2 : register(t2);
+StructuredBuffer<float4> Wall3 : register(t3);
+
+RWStructuredBuffer<float> Result1D : register(u4);
+RWStructuredBuffer<float2> Result2D : register(u5);
+// float4 for simplification in memory alignment
+RWStructuredBuffer<float4> Result3D : register(u6);
+RWStructuredBuffer<float4> Result4D : register(u7);
+
+[numthreads(1,1,1)]
+void main() {
+  // scenario 1: -----> |
+
+  float Wall1_1D = reflect(IncidentRay[0].x, Wall1[0].x);
+  Result1D[0] = Wall1_1D;
+  float2 Wall1_2D = reflect(IncidentRay[0].xy, normalize(Wall1[0].xy));
+  Result2D[0] = Wall1_2D;
+  float3 Wall1_3D = reflect(IncidentRay[0].xyz, normalize(Wall1[0].xyz));
+  Result3D[0] = float4(Wall1_3D, 0.0);
+  float4 Wall1_4D = reflect(IncidentRay[0].xyzw, normalize(Wall1[0].xyzw));
+  Result4D[0] = Wall1_4D;
+
+  // scenario 2: -----> /
+
+  // The 1D scenario isn't applicable because there is no slanted
+  // wall that can be normalized in 1 dimensional space, since a slant
+  // requires at least 2 dimensions.
+  Result1D[1] = 0.0; 
+
+  float2 Wall2_2D = reflect(IncidentRay[0].xy, normalize(Wall2[0].xy));
+  Result2D[1] = Wall2_2D;
+  float3 Wall2_3D = reflect(IncidentRay[0].xyz, normalize(Wall2[0].xyz));
+  Result3D[1] = float4(Wall2_3D, 0.0);
+  float4 Wall2_4D = reflect(IncidentRay[0].xyzw, normalize(Wall2[0].xyzw));
+  Result4D[1] = Wall2_4D;
+
+  // scenario 3: -----> -
+
+  // The 1D scenario isn't applicable because there is no horizontal
+  // wall that can be normalized in 1 dimensional space, since the x
+  // component must be 0.
+  float Wall3_1D = 0.0;
+  Result1D[2] = Wall3_1D;
+  float2 Wall3_2D = reflect(IncidentRay[0].xy, normalize(Wall3[0].xy));
+  Result2D[2] = Wall3_2D;
+  float3 Wall3_3D = reflect(IncidentRay[0].xyz, normalize(Wall3[0].xyz));
+  Result3D[2] = float4(Wall3_3D, 0.0);
+  float4 Wall3_4D = reflect(IncidentRay[0].xyzw, normalize(Wall3[0].xyzw));
+  Result4D[2] = Wall3_4D;
+}
+
+//--- pipeline.yaml
+
+---
+Shaders:
+  - Stage: Compute
+    Entry: main
+    DispatchSize: [1, 1, 1]
+Buffers:
+  - Name: IncidentRay
+    Format: Float32
+    Stride: 16
+    Data: [ 2.5, -0.0, 0.0, 0.0 ]
+  - Name: Wall1
+    Format: Float32
+    Stride: 16
+    Data: [ -1.0, 0.0, 0.0, 0.0 ]

```

Comment:
Using 1.0 and so many zero components means this is only testing a degenerate edge case.  I really think we should be testing with combinations of numbers in reasonable ranges and using a ULP tolerance at least as high as the potential variance in expansions and operation precisions.